Complete first sprint. Step second

uncledeni opened the pull request:
Complete first sprint
Ask you feedback on it

uncledeni committed:
First step compelete

Add MainPage component
Add BurgerConstructor component
Add BurgerIngredients component
Add AppHeader component

uncledeni committed:
fixes according to pull request

sematic tags usage update;
Change all '.css' files usage to 'module.css'; update classNames;
fix 'app' file; update import;

uncledeni committed:
Coplete first sprint. Step second

install 'prop-types' npm;
add PropTypes check;
add "react-modals" div for out-root rendering;
create api for dynamic data collecting; use fetch and try/catch;
create Modal component template; create ModalOverlay component;
create OrderDetails modal;
create ingredientDetails modal;

gennady-bars commented on src/shared/components/modal/modal.jsx:
```diff
@@ -0,0 +1,47 @@
+const modalRoot = document.getElementById("react-modals");
+
+export const Modal = (props) => {
+    if (props.isOpen) {
+        document.addEventListener('keydown', function (event) {
+            if (event.key === "Escape") props.handlerOpen(false);
+        });
+    }
+
+    if (props.isOpen) {
+        document.removeEventListener('keydown', function (event) {
+            if (event.key === "Escape") props.handlerOpen(false);
+        });
```
Принято удалять обработчик в `clean up`-функции хука `useEffect` (даже если компонент не будет размонтирован). Вдруг, поменяете что-то в логике, и он все-таки начнет размонтироваться.


И нужно правильно удалять обработчик

В таком варианте невозможно удалить обработчик, так как Вы его **ссылку** не передаете. Там Вы передаете **анонимную** стрелочную функцию (у неё нет постоянной ссылки). Движок браузера не сможет найти по ссылке ее и удалить. Ошибки никакой браузер не выдаст, так как ему без разницы, что он не нашел функцию.

gennady-bars commented on src/shared/components/modal/modal.jsx:
```diff
@@ -0,0 +1,47 @@
+
+export const Modal = (props) => {
+    if (props.isOpen) {
+        document.addEventListener('keydown', function (event) {
```
Можно добавить обработчик `Escape` 

```jsx

  useEffect(() => {
    function closeByEscape(evt) {
      if(evt.key === 'Escape') {
        closeModal();
      }
    }
      document.addEventListener('keydown', closeByEscape);
      return () => {
        document.removeEventListener('keydown', closeByEscape);
      }

  }, []) 

```

Объявляем функцию внутри `useEffect`, чтобы она не теряла свою ссылку при обновлении компонента.
И не забываем удалять обработчик в `clean up` функции через `return`
А также массив зависимостей

gennady-bars commented on src/shared/api/get-data-service.jsx:
```diff
@@ -0,0 +1,11 @@
+
+export const GetDataService = async () => {
+    try {
+        let response = await fetch(BURGER_DATA);
+        let json = await response.json();
```
Любой ответ от сервера нужно проверять на `ok` (на **корректность**) и вызывать ошибку, если ответ не `ok`. Таким образом, в `catch` Вы будете ловить и ошибки сервера, и ошибки невалидных данных.


```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Ошибка ${res.status}`);
            })

```

gennady-bars replied:
Если переменная объявлена через `let`, её значение должно быть напрямую перезаписано.
Если переменная **не переназначается**, то она объявляется через `const`.


Все переменные нужно объявлять по возможности через `const`.
В больших функциях бывает так, что Вы можете не заметить, если переменная переназначится, и это может оказаться причиной багов. Особенно в замыканиях. 
`const` дает Вам уверенность в том, что Вы всегда будете видеть одно и то же значение.

gennady-bars commented on src/shared/api/get-data-service.jsx:
```diff
@@ -0,0 +1,11 @@
+        let response = await fetch(BURGER_DATA);
+        let json = await response.json();
+        return json;
+    } catch (err) {
+        alert("Ошибка HTTP: " + err.status);
```
Не стоит использовать `catch` в `Api`, так как ловить ошибки нужно в самом конце цепочки запроса, а конец цепочки находится в  (`Ingredients`) после всего  кода, который происходит в `then`. И в реальных проектах в `catch` происходит много логики: открывается попап с ошибкой и тд. Поэтому файл `api` не может знать, что делать с ошибкой внутри.

gennady-bars commented on src/widgets/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,132 @@
+
+    useEffect(() => {
+        const getData = async () => {
+            const res = await GetDataService();
```
`GetDataService`

Любые переменные в проекте нужно называть в `camelCase` по чек-листу. Это значит, что они должны быть с **маленькой буквы**, если это не классы или компоненты реакта

gennady-bars commented on src/widgets/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,132 @@
+        const getData = async () => {
+            const res = await GetDataService();
+            setIngredientsData(res.data);
+        }
```
Для любых запросов на сервер в **самом конце цепочки** запроса нужно осуществлять проверку на возможные ошибки с помощью `catch`.

gennady-bars commented on src/widgets/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,132 @@
+
+const Ingredients = () => {
+    const [ingredientsData, setIngredientsData] = useState([]);
+    const [openModal, setOpenModal] = useState(false);
```
**Можно лучше**

Если интересно, можно создать **кастомный хук** для управления модальным окном:

```js
import { useState, useCallback } from "react";

// кастомные хуки всегда должны начинаться с глагола `use`, чтобы реакт понял, что это хук. Он следит за их вызовами
export const useModal = () => {
  const [isModalOpen, setIsModalOpen] = useState(false);

  // `useCallback` нужен для того, чтобы зафиксировать ссылку на функцию. Таким образом уменьшится кол-во перерисовок компонента, куда будет передана эта функция
  const openModal = useCallback(() => {
    setIsModalOpen(true);
  }, []);

  const closeModal = useCallback(() => {
    setIsModalOpen(false);
  }, []);

  return {
    isModalOpen,
    openModal,
    closeModal,
  };
};

```

Этот код помещают в отдельный файл `useModal.js` в папке `hooks` и импортируют функцию туда, где нужно создать показатели открытости/закрытости попапа (у каждого попапа они свои)

И теперь нужно просто вызывать хук в том компоненте, где нужно управлять попапом. Всего одна строчка кода будет (под каждый попап своя):

```js
  const { isModalOpen, openModal, closeModal } = useModal();

````

uncledeni replied:
спасибо за рекомендацию и подсказку, такой подход действительно куда лаконичнее и удобнее

gennady-bars commented on src/widgets/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,132 @@
+            <IngredientsStack openModal={openModal} setOpenModal={setOpenModal} setIngredient={setIngredient} title="Булки" data={ingredientsData} type={BUN_TYPE} />
+            <IngredientsStack openModal={openModal} setOpenModal={setOpenModal} setIngredient={setIngredient} title="Начинка" data={ingredientsData} type={MAIN_TYPE} />
+            <IngredientsStack openModal={openModal} setOpenModal={setOpenModal} setIngredient={setIngredient} title="Соусы" data={ingredientsData} type={SAUCE_TYPE} />
+            <Modal isOpen={openModal} handlerOpen={setOpenModal}>
```
Не нужно передавать в пропсы `Modal` показатель открытости `isOpen`, если не делаете **плавное** открытие попапов. Это усложняет логику внутри и усложняет правильное удаление обработчика `Escape`.


Вот так целиком будет появляться попап

```jsx

      {openModal && (
        <Modal handlerOpen={setOpenModal}>
          <IngredientDetails
            image_large={ingredient.image_large}
            name={ingredient.name}
            calories={ingredient.calories}
            proteins={ingredient.proteins}
            fat={ingredient.fat}
            carbohydrates={ingredient.carbohydrates}
          />
        </Modal>
      )}
```

gennady-bars commented on src/widgets/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,132 @@
+            <IngredientsStack openModal={openModal} setOpenModal={setOpenModal} setIngredient={setIngredient} title="Соусы" data={ingredientsData} type={SAUCE_TYPE} />
+            <Modal isOpen={openModal} handlerOpen={setOpenModal}>
+                <IngredientDetails
+                    image_large={ingredient.image_large}
+                    name={ingredient.name}
+                    calories={ingredient.calories}
+                    proteins={ingredient.proteins}
+                    fat={ingredient.fat}
+                    carbohydrates={ingredient.carbohydrates}
```
**Можно лучше**

* Лучше всего передавать **объект ингредиента** целиком в пропсы под понятным названием: `ingredient={ingredient}`, чтобы его потом было удобно типизировать единым универсальным типом любого ингредиента

gennady-bars commented on src/widgets/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,132 @@
+
+IngredientElem.propTypes = {
+    ingredient: PropTypes.shape({
+        calories: PropTypes.number,
+        carbohydrates: PropTypes.number,
+        fat: PropTypes.number,
+        image: PropTypes.string,
+        image_large: PropTypes.string,
+        image_mobile: PropTypes.string,
+        name: PropTypes.string,
+        price: PropTypes.number,
+        proteins: PropTypes.number,
+        type: PropTypes.string,
+        __v: PropTypes.number,
+        _id: PropTypes.string,
```
Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `ingredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on src/widgets/burger-ingredients/components/ingredient-details/ingredient-details.jsx:
```diff
@@ -0,0 +1,43 @@
+}
+
+IngredientDetails.propTypes = {
+    image_large: PropTypes.string,
+    name: PropTypes.string,
+    calories: PropTypes.number,
+    proteins: PropTypes.number,
+    fat: PropTypes.number,
+    carbohydrates: PropTypes.number
```
Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `ingredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно

gennady-bars commented on src/shared/components/modal-overlay/modal-overlay.jsx:
```diff
@@ -0,0 +1,17 @@
+    console.log(props)
+    return (
+        <div onClick={() => {
+            props.handlerOpen(false);
```
`props.handlerOpen(false)` дублируется при закрытии попапа, значит, нужно объявить функцию `closeModal` и передавать ее в пропсы `Modal`, а потом из `Modal` в `ModalOverlay`

gennady-bars commented on src/shared/components/modal/modal.jsx:
```diff
@@ -0,0 +1,47 @@
+export const Modal = (props) => {
+    if (props.isOpen) {
+        document.addEventListener('keydown', function (event) {
+            if (event.key === "Escape") props.handlerOpen(false);
```
`props.handlerOpen(false)` дублируется при закрытии попапа, значит, нужно объявить функцию `closeModal` и передавать ее в пропсы `Modal`, а потом из `Modal` в `ModalOverlay`

gennady-bars commented on src/shared/components/modal/modal.jsx:
```diff
@@ -0,0 +1,47 @@
+                            {props.children}
+                            <span className={`${ModalStyles.closeIcon}`}>
+                                <CloseIcon onClick={() => {
+                                    props.handlerOpen(false);
```
`props.handlerOpen(false)` дублируется при закрытии попапа, значит, нужно объявить функцию `closeModal` и передавать ее в пропсы `Modal`, а потом из `Modal` в `ModalOverlay`

gennady-bars commented on src/widgets/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,132 @@
+            setIngredientsData(res.data);
+        }
+        getData();
+    }, [])
```
Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Отлично, что используете пустой массив зависимостей `[]`, чтобы `useEffect` сработал 1 раз


но есть некоторые недочеты:


* Принято удалять обработчик в `clean up`-функции хука `useEffect` (даже если компонент не будет размонтирован). Вдруг, поменяете что-то в логике, и он все-таки начнет размонтироваться.
* Любой ответ от сервера нужно проверять на `ok` (на **корректность**) и вызывать ошибку, если ответ не `ok`. Таким образом, в `catch` Вы будете ловить и ошибки сервера, и ошибки невалидных данных.
* Если переменная **не переназначается**, то она объявляется через `const`.
* Не стоит использовать `catch` в `Api`, так как ловить ошибки нужно в самом конце цепочки запроса, а конец цепочки находится в  (`Ingredients`) после всего  кода, который происходит в `then`. И в реальных проектах в `catch` происходит много логики: открывается попап с ошибкой и тд. Поэтому файл `api` не может знать, что делать с ошибкой внутри.
* Любые переменные в проекте нужно называть в `camelCase` по чек-листу. Это значит, что они должны быть с **маленькой буквы**, если это не классы или компоненты реакта
* Для любых запросов на сервер в **самом конце цепочки** запроса нужно осуществлять проверку на возможные ошибки с помощью `catch`.
* Не нужно передавать в пропсы `Modal` показатель открытости `isOpen`, если не делаете **плавное** открытие попапов. Это усложняет логику внутри и усложняет правильное удаление обработчика `Escape`.
* Тип для ингредиента (propTypes) повторяется в некоторых компонентах. Не нужно его копировать. Это очень сложный объект. Нужно вынести его в константу `ingredientType` в отдельный файл `types.js` в папку `utils` и импортировать туда, где нужно
* `props.handlerOpen(false)` дублируется при закрытии попапа, значит, нужно объявить функцию `closeModal` и передавать ее в пропсы `Modal`, а потом из `Modal` в `ModalOverlay`

**Можно лучше**

* Лучше всего передавать **объект ингредиента** целиком в пропсы под понятным названием: `ingredient={ingredient}`, чтобы его потом было удобно типизировать единым универсальным типом любого ингредиента

uncledeni committed:
fixes according to pull request

create 'types.js'; add all complex type in it;
create custom hook 'useModal'; rework modal open/close functional;
add 'useModal' into 'burger-ingredients' and 'burger-constructor';
rework server request handler; add try/catch to 'burger-ingredients' as last request border;
rework 'ingredient-details' props;

gennady-bars commented on src/widgets/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,95 @@
+        <div className={BurgerConstructorStyles.burgerConstructorWrapper}>
+            <ConstructorComponent />
+            <Info constructorModal={openModal} />
+            {isModalOpen && <Modal handlerOpen={closeModal}>
```
`handlerOpen` теперь нужно назвать просто `onClose` или `handleClose` или `closeModal`, так как 100% закрывает попап, а не открывает или еще что-то в этом роде

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
